src/Features/CSharp/Portable/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.Rewriter.cs

```diff
@@ -0,0 +1,263 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp.ConvertSwitchStatementToExpression
+{
+    using static SyntaxFactory;
+
+    internal sealed partial class ConvertSwitchStatementToExpressionCodeFixProvider
+    {
+        private sealed class Rewriter : CSharpSyntaxVisitor<ExpressionSyntax>
+        {
+            private readonly ArrayBuilder<ExpressionSyntax> _assignmentTargets;
+
+            private Rewriter()
+            {
+                _assignmentTargets = ArrayBuilder<ExpressionSyntax>.GetInstance();
+            }
+
+            private void Free()
+            {
+                _assignmentTargets.Free();
+            }
+
+            public static StatementSyntax Rewrite(SwitchStatementSyntax node, SemanticModel semanticModel, SyntaxEditor editor)
+            {
+                var rewriter = new Rewriter();
+                var switchExpression = rewriter.VisitSwitchStatement(node);
+                var generateDeclaration = rewriter.TryRemoveVariableDeclarators(node, semanticModel, editor);
+                var finalStatement = rewriter.GetFinalStatement(switchExpression, generateDeclaration);
+                rewriter.Free();
+                return finalStatement;
+            }
+
+            private bool TryRemoveVariableDeclarators(SwitchStatementSyntax node, SemanticModel semanticModel, SyntaxEditor editor)
+            {
+                // Try to remove the varaiable declarator only if these are simple identifiers.
+                if (_assignmentTargets.Count == 0 ||
+                    !_assignmentTargets.All(target => target.IsKind(SyntaxKind.IdentifierName)))
+                {
+                    return false;
+                }
+
+                var symbols = _assignmentTargets.Select(target => semanticModel.GetSymbolInfo(target).Symbol);
+                // If all variables are local and data does not flows in for any of them, 
+                // we can assign the switch expression directly to a variable declaration.
+                foreach (var symbol in symbols)
+                {
+                    if (symbol == null)
+                    {
+                        return false;
+                    }
+
+                    if (symbol.Kind != SymbolKind.Local)
+                    {
+                        return false;
+                    }
+
+                    var syntaxReferences = symbol.DeclaringSyntaxReferences;
+                    if (syntaxReferences.Length != 1)
+                    {
+                        return false;
+                    }
+
+                    var declarator = (VariableDeclaratorSyntax)syntaxReferences[0].GetSyntax();
+                    if (declarator.Initializer != null)
+                    {
+                        return false;
+                    }
+
+                    var dataFlow = semanticModel.AnalyzeDataFlow(
+                        declarator.GetAncestor<StatementSyntax>(),
+                        node.GetPreviousStatement());
```

Comment:
worth expalining the .GetPreviousStatement.

Also, is it required that declarator's statement's parent and the previous statemnet of the switch need to be in the same block?